File: flink-sql/src/main/java/org/apache/flink/table/plan/rules/DynamicStreamTableSourceScanRule.java
package org.apache.flink.table.plan.rules;

import org.apache.calcite.plan.RelOptRuleCall;
import org.apache.calcite.plan.RelTraitSet;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.convert.ConverterRule;
import org.apache.flink.table.plan.nodes.DynamicFlinkConventions;
import org.apache.flink.table.plan.nodes.FlinkConventions;
import org.apache.flink.table.plan.nodes.datastream.DynamicTableSourceScan;
import org.apache.flink.table.plan.nodes.logical.FlinkLogicalTableSourceScan;
import org.apache.flink.table.sources.DynamicStreamTableSource;
import org.apache.flink.table.sources.TableSource;

public class DynamicStreamTableSourceScanRule extends ConverterRule {

  public static DynamicStreamTableSourceScanRule INSTANCE = new DynamicStreamTableSourceScanRule();

  private DynamicStreamTableSourceScanRule() {
    super(FlinkLogicalTableSourceScan.class, FlinkConventions.LOGICAL(), DynamicFlinkConventions.DYNAMIC_DATA_STREAM, "DynamicStreamTableSourceScanRule");
  }

  @Override
  public boolean matches(RelOptRuleCall call) {
    FlinkLogicalTableSourceScan scan = call.rel(0);

    TableSource<?> tableSource = scan.tableSource();

    return tableSource instanceof DynamicStreamTableSource;
  }

  @Override
  public RelNode convert(RelNode rel) {
    FlinkLogicalTableSourceScan scan = (FlinkLogicalTableSourceScan) rel;
    RelTraitSet traitSet = rel.getTraitSet().replace(DynamicFlinkConventions.DYNAMIC_DATA_STREAM);

    return new DynamicTableSourceScan(rel.getCluster(), traitSet, scan.getTable(),
        (DynamicStreamTableSource) scan.tableSource(), scan.selectedFields());
  }
}

File: flink-extended-sql/src/main/scala/org/apache/flink/table/plan/rules/datastream/DynamicStreamTableSourceScanRule.scala
package org.apache.flink.table.plan.rules.datastream

import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelTraitSet}
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.convert.ConverterRule
import org.apache.calcite.rel.core.TableScan
import org.apache.flink.table.plan.nodes.FlinkConventions
import org.apache.flink.table.plan.nodes.datastream.DynamicStreamTableSourceScan
import org.apache.flink.table.plan.nodes.logical.FlinkLogicalTableSourceScan
import org.apache.flink.table.plan.schema.TableSourceTable
import org.apache.flink.table.sources.StreamTableSource

class DynamicStreamTableSourceScanRule
  extends ConverterRule(
    classOf[FlinkLogicalTableSourceScan],
    FlinkConventions.LOGICAL,
    FlinkConventions.DATASTREAM,
    "DynamicStreamTableSourceScanRule") {

  override def matches(call: RelOptRuleCall): Boolean = {
    val scan: TableScan = call.rel(0).asInstanceOf[TableScan]

    val sourceTable = scan.getTable.unwrap(classOf[TableSourceTable[_]])
    sourceTable != null && sourceTable.isStreamingMode
  }

  override def convert(rel: RelNode): RelNode = {
    val scan: FlinkLogicalTableSourceScan = rel.asInstanceOf[FlinkLogicalTableSourceScan]
    val traitSet: RelTraitSet = rel.getTraitSet.replace(FlinkConventions.DATASTREAM)

    new DynamicStreamTableSourceScan(
      rel.getCluster,
      traitSet,
      scan.getTable,
      scan.tableSchema,
      scan.tableSource.asInstanceOf[StreamTableSource[_]],
      scan.selectedFields
    )
  }

}

object DynamicStreamTableSourceScanRule {

  val INSTANCE: RelOptRule = new DynamicStreamTableSourceScanRule

}
